<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysing transcript 5’-profiling data using icetea • icetea</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><meta property="og:title" content="Analysing transcript 5’-profiling data using icetea">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">icetea</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/mapcap_analysis.html">Analysing transcript 5'-profiling data using icetea</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/vivekbhr/icetea">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Analysing transcript 5’-profiling data using icetea</h1>
                        <h4 class="author">Vivek Bhardwaj and Thomas Manke</h4>
            <address class="author_afil">
      Max Planck Institute of Immunobiology and Epigenetics, Stübeweg 51, 79108, Freiburg, Germany<br><h4 class="date">03/05/2018</h4>
      
      <small>Source: <a href="https://github.com/vivekbhr/icetea/blob/master/vignettes/mapcap_analysis.Rmd"><code>vignettes/mapcap_analysis.Rmd</code></a></small>
      <div class="hidden name"><code>mapcap_analysis.Rmd</code></div>

    </address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Transcript 5’-profiling techniques detect 5’ sites of the expressed transcripts in the genome, allowing us to detect transcription start sites in the genome at high resolution. <em>CAGE</em> (Cap Analysis of Gene Expression) is one such technique. Programs for analysis of CAGE data use the clustering of CAGE tags (CTSS) to annotate transcription start sites and detect promoter shifts. Recently, faster and more efficient 5’-profiling techniques have been developed which allow early sample multiplexing, removal of PCR duplicates using random barcodes, and use paired-end sequencing. <strong>icetea</strong> can be used to process such data. It also implements a new method for detection of transcription start sites based on local enrichment of genomic windows, that can take advantage of biological replicates. It further allows users to perform differential expression analysis of transcription start sites between groups of samples, therefore integrating transcript annotation and expression analysis. While icetea analysis is mostly useful for multiplexed, paired-end experiments with replicates, conventional CAGE data can also be analysed. <em>icetea</em> analysis compliments other programs that perform promoter profiling and expression analysis based on clusteing of CAGE tags.</p>
    </div>
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>Here we show the usability of <em>icetea</em> in the analysis of data from a recently developed paired-end, multiplexed 5’-profiling protocol called MAPCap. <strong>MAPCap</strong> (Multiplexed Affinity Purification of Capped RNA) allows fast and accutate detection of transcription start sites and expression analysis of multiplexed, low-input RNA samples.</p>
<p><a name="top"></a></p>
</div>
<div id="quick-start" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick Start</h2>
<p>Below is the set of minimal steps for the processing of MAPCap data for the detection of TSS. Starting from the raw fastq files, we can either perform quality trimming using a standard program (eg. <a href="https://cutadapt.readthedocs.io/en/stable/">cutadapt</a>/<a href="https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/">trimgalore</a>) or simply begin the analysis by creating a <code>CapSet</code> object. The minimal information required to create a <code>CapSet</code> object is the path to raw/trimmed FASTQ files and a vector of sample names. For multiplexed FASTQ files, additionally a vector of sample indexes should be provided to proceed with de-multiplexing.</p>
<p>The following steps creates the object, demultiplexes the fastq, maps them, filters them and detects the transcription start sites.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load the package </span>
<span class="kw">library</span>(icetea)

<span class="co"># provide demultiplexing barcodes and sample names</span>
idxlist &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"CAAGTG"</span>, <span class="st">"TTAGCC"</span>, <span class="st">"GTGGAA"</span>, <span class="st">"TGTGAG"</span>)
dir &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="dt">package=</span><span class="st">"icetea"</span>)
<span class="co"># corresponding sample names</span>
fnames &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"embryo1"</span>, <span class="st">"embryo2"</span>, <span class="st">"embryo3"</span>, <span class="st">"embryo4"</span>)
## create CapSet object
cs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/newCapSet.html">newCapSet</a></span>(<span class="dt">expMethod =</span> <span class="st">'MAPCap'</span>,
                <span class="dt">fastq_R1 =</span> <span class="kw">file.path</span>(dir, <span class="st">'mapcap_test_R1.fastq.gz'</span>),
                <span class="dt">fastq_R2 =</span> <span class="kw">file.path</span>(dir, <span class="st">'mapcap_test_R2.fastq.gz'</span>),
                <span class="dt">idxList =</span> idxlist,
                <span class="dt">sampleNames =</span> fnames)

<span class="co"># demultiplex fastq and trim the barcodes</span>
<span class="kw">dir.create</span>(<span class="st">"splitting"</span>)
cs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/demultiplexFASTQ.html">demultiplexFASTQ</a></span>(cs, <span class="dt">max_mismatch =</span> <span class="dv">1</span>, <span class="dt">outdir =</span> <span class="st">"splitting"</span>, <span class="dt">ncores =</span> <span class="dv">10</span>)

<span class="co"># map fastq</span>
<span class="kw">dir.create</span>(<span class="st">"mapping"</span>)
cs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mapCaps.html">mapCaps</a></span>(cs, subread_idx, <span class="dt">outdir =</span> <span class="st">"mapping"</span>, <span class="dt">ncores =</span> <span class="dv">20</span>, <span class="dt">logfile =</span> <span class="st">"mapping/subread_mapping.log"</span>)

<span class="co"># filter PCR duplicates</span>
<span class="kw">dir.create</span>(<span class="st">"removedup"</span>)
cs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/filterDuplicates.html">filterDuplicates</a></span>(cs, <span class="dt">outdir =</span> <span class="st">"removedup"</span>)

<span class="co"># detect TSS</span>
<span class="kw">dir.create</span>(<span class="st">"tssCalling"</span>)
cs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/detectTSS.html">detectTSS</a></span>(cs, <span class="dt">groups =</span> <span class="kw">c</span>(<span class="st">"wt"</span>, <span class="st">"wt"</span>, <span class="st">"mut"</span>, <span class="st">"mut"</span>), <span class="dt">outfile_prefix =</span> <span class="st">"tssCalling/testTSS"</span>)</code></pre></div>
</div>
<div id="help-and-citations" class="section level2">
<h2 class="hasAnchor">
<a href="#help-and-citations" class="anchor"></a>Help and citations</h2>
<div id="how-to-get-help" class="section level3">
<h3 class="hasAnchor">
<a href="#how-to-get-help" class="anchor"></a>How to get help</h3>
<p>ICETEA questions could be posted to the <a href="https://support.bioconductor.org">Bioconductor support site</a>, which serves as a searchable knowledge base of questions and answers.</p>
<p>Posting a question and tagging with “icetea” will automatically send an alert to the package authors to respond on the support site. See the first question in the list of Frequently Asked Questions (FAQ) for information about how to construct an informative post.</p>
</div>
<div id="how-to-cite-icetea" class="section level3">
<h3 class="hasAnchor">
<a href="#how-to-cite-icetea" class="anchor"></a>How to cite ICETEA</h3>
<p>To cite <em>icetea</em> package in your analysis type <code>citation("icetea")</code> from within your R session.</p>
<p><a name="capset"></a></p>
</div>
</div>
<div id="the-capset-object" class="section level2">
<h2 class="hasAnchor">
<a href="#the-capset-object" class="anchor"></a>The CapSet object</h2>
<p>The workflow begins by creating an object of class <code>CapSet</code> using the function <code>newCapSet</code>. This object contains information about the experiment method (CAGE <span class="citation">(Kodzius et al. 2006)</span>, RAMPAGE <span class="citation">(Batut et al. 2013)</span> or MAPCap), along with the path of the multiplexed fastq files. To allow de-multiplexing as , the demultiplexing barcodes should be provided to <code>idxList</code> and the corresponding samplenames should be provided to <code>sampleNames</code>.</p>
<p>With this information, create the CapSet object as follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># provide demultiplexing barcodes as strings</span>
idxlist &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"CAAGTG"</span>, <span class="st">"TTAGCC"</span>, <span class="st">"GTGGAA"</span>, <span class="st">"TGTGAG"</span>)

<span class="co"># provide corresponding sample names</span>
fnames &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"embryo1"</span>, <span class="st">"embryo2"</span>, <span class="st">"embryo3"</span>, <span class="st">"embryo4"</span>)

<span class="co"># `dir` contains example data provided with this package</span>
dir &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="dt">package =</span> <span class="st">"icetea"</span>)

## CapSet object from raw (multiplexed) fastq files
<span class="kw">library</span>(icetea)
cs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/newCapSet.html">newCapSet</a></span>(<span class="dt">expMethod =</span> <span class="st">'MAPCap'</span>,
                <span class="dt">fastq_R1 =</span> <span class="kw">file.path</span>(dir, <span class="st">'mapcap_test_R1.fastq.gz'</span>),
                <span class="dt">fastq_R2 =</span> <span class="kw">file.path</span>(dir, <span class="st">'mapcap_test_R2.fastq.gz'</span>),
                <span class="dt">idxList =</span> idxlist,
                <span class="dt">sampleNames =</span> fnames)</code></pre></div>
<p>The <code>CapSet</code> object is designed to allow the steps in the analysis to proceed like a pipeline. It holds the information on the output files from each step, which are then passed on to the next step. This information is stored in a DataFrame object, which can be retreived and manipulated using the function <code>sampleInfo</code>. In case of underlying files being deleted or moved to another folder, we can reset the sample information this way.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">si &lt;-<span class="st"> </span><span class="kw">sampleInfo</span>(cs)
dir &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/bam"</span>, <span class="dt">package =</span> <span class="st">"icetea"</span>)
si<span class="op">$</span>mapped_file &lt;-<span class="st"> </span><span class="kw">list.files</span>(dir, <span class="dt">pattern =</span> <span class="st">".bam$"</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)
<span class="kw">sampleInfo</span>(cs) &lt;-<span class="st"> </span>si</code></pre></div>
<p>The <code>CapSet</code> object doesn’t need to be created at the beginning of the workflow. Users can perform the initial, computationally expensive steps of FASTQ de-multiplexing, mapping and PCR de-duplication outside of the R environment. The <code>CapSet</code> object can be created at any point of the workflow, using the output files from current step.</p>
<p>After mapping and de-duplication :</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dir &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/filtered_bam"</span>, <span class="dt">package =</span> <span class="st">"icetea"</span>)
cs2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/newCapSet.html">newCapSet</a></span>(<span class="dt">expMethod =</span> <span class="st">'MAPCap'</span>,
                 <span class="dt">filtered_file =</span> <span class="kw">list.files</span>(dir, <span class="dt">pattern =</span> <span class="st">".bam$"</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>),
                 <span class="dt">sampleNames =</span> fnames)
<span class="co">#&gt; Checking de-duplicated file</span></code></pre></div>
<p>Providing output files from previous steps are not necessory, but providing them allows calculating additional metadata (like number of mapped reads) which could be useful for plotting (see section <a href="#QC">QC</a> )</p>
</div>
<div id="analysis-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#analysis-workflow" class="anchor"></a>Analysis workflow</h2>
<p><a name="demult"></a></p>
<div id="de-multiplexing-the-fastq" class="section level3">
<h3 class="hasAnchor">
<a href="#de-multiplexing-the-fastq" class="anchor"></a>De-multiplexing the fastq</h3>
<p>Experiments like MAPCap and RAMPAGE <span class="citation">(Batut et al. 2013)</span> produce multiplexed fastq files with sample indicies and PCR barcodes attached to the fastq sequence. The tool <code>demultiplexFASTQ</code> de-multiplexes the samples using the sample barcode, producing fastq files corresponding to each sample. It also trims off these barcodes and attaches them in the header of the fastq files for further processing.</p>
<p>For RAMPAGE <span class="citation">(Batut et al. 2013)</span> protocol, the modified header looks like this : <code>&lt;read ID&gt;#&lt;sample barcode&gt;:&lt;pseudo-random barcode&gt;</code> For MAPCap protocol, the modified header looks like this : <code>&lt;read ID&gt;#&lt;sample barcode&gt;:&lt;random barcode&gt;:&lt;replicate barcode&gt;</code> For conventional CAGE <span class="citation">(Kodzius et al. 2006)</span> protocols, we assume that standard Illumina de-multiplexing has already been performed and this step is not required.</p>
<p>Function returns a modified CapSet object that contains location of demultiplexed files along with processing statistics.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># demultiplex fastq and trim the barcodes</span>
<span class="kw">dir.create</span>(<span class="st">"splitting"</span>)
<span class="co">#&gt; Warning in dir.create("splitting"): 'splitting' already exists</span>
cs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/demultiplexFASTQ.html">demultiplexFASTQ</a></span>(cs, <span class="dt">max_mismatch =</span> <span class="dv">1</span>, <span class="dt">outdir =</span> <span class="st">"splitting"</span>, <span class="dt">ncores =</span> <span class="dv">10</span>)
<span class="co">#&gt; de-multiplexing the FASTQ file</span></code></pre></div>
<p>It takes about 6 min (340 sec) to trim and de-multiplex 1M PE reads into 12 sampels (12 pair of fastqs), using 1 thread. This can be done under 45 seconds if 10 threads are used.</p>
<p>Optionally, we can skip the above process and perform post-mapping de-multiplexing on the BAM files (see below).</p>
</div>
<div id="mapping-the-fastqs" class="section level3">
<h3 class="hasAnchor">
<a href="#mapping-the-fastqs" class="anchor"></a>Mapping the fastqs</h3>
<p>The demultiplexed fastqs can now be mapped using the <code>mapCaps</code> function. This function is a wrapper over the <code>subjunc</code> function from <code>Rsubread</code> package <span class="citation">(Liao, Smyth, and Shi 2013)</span>. It additionally performs sorting and collects mapping statistics of the mapped files, stored in the modified <code>CapSet</code> object.</p>
<p>In order to run the function we first create a subread index of our genome.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dir.create</span>(<span class="st">"genome_index"</span>)
<span class="kw">library</span>(Rsubread)
<span class="kw"><a href="http://www.rdocumentation.org/packages/Rsubread/topics/buildindex">buildindex</a></span>(<span class="dt">basename =</span> <span class="st">"genome_index/dm6"</span>, <span class="dt">reference =</span> <span class="st">"/path/to/dm6/genome.fa"</span>)</code></pre></div>
<p>We can now perform the mapping.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># provide location of a subread index file</span>
subread_idx &lt;-<span class="st"> "genome_index/dm6"</span>
<span class="co"># map fastq</span>
cs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mapCaps.html">mapCaps</a></span>(cs, subread_idx, <span class="dt">outdir =</span> <span class="st">"mapping"</span>, <span class="dt">ncores =</span> <span class="dv">20</span>, <span class="dt">logfile =</span> <span class="st">"mapping/subread_mapping.log"</span>)
<span class="co"># you can save the CapSet object for later use</span>
<span class="kw">save</span>(cs, <span class="dt">file =</span> <span class="st">"myCapSet.Rdata"</span>)</code></pre></div>
<p><strong>Note:</strong> The package <code>Rsubread</code> is not available for windows. Windows users would need to map their demultiplexed files using another tool.</p>
<p><strong>Note:</strong> Optionally, we can perform mapping directly on the multiplexed files and perform the de-multiplexing afterwards (see below). This option is only recommended where the FASTQ files are not very large (for example, where overall sequencing depth is low, below 10 Mil).</p>
</div>
<div id="optional-post-mapping-de-multiplexing" class="section level3">
<h3 class="hasAnchor">
<a href="#optional-post-mapping-de-multiplexing" class="anchor"></a>(optional) Post mapping de-multiplexing</h3>
<p>Demultiplxeing can be performed on the BAM files after mapping, using the <code>splitBAM_byIndex</code> and (for MAPCap) <code>splitBAM_byRepindex</code> functions. These functions assume that the reads have been trimmed and the sample barcodes have been attached to the header in the way as <a href="#demult">described above</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/splitBAM_byIndex.html">splitBAM_byIndex</a></span>(bamFile, index_list, outfile_list, <span class="dt">max_mismatch =</span> <span class="dv">1</span>, <span class="dt">ncores =</span> <span class="dv">10</span>)</code></pre></div>
</div>
<div id="filtering-pcr-duplicates" class="section level3">
<h3 class="hasAnchor">
<a href="#filtering-pcr-duplicates" class="anchor"></a>Filtering PCR duplicates</h3>
<p>Experiments like MAPCap and RAMPAGE provide us a way to remove sequencing reads which are PCR duplicates from the mapped data. Random barcodes present in the read sequence are used for this purpose. In MAPCap, pre-designed random barcodes present in the oligos serve as UMIs, while in RAMPAGE, the sequences used as RT-PCR primers are treated as pseudo-random barcodes. PCR duplicates are recognized as reads that map to the same 5’ position and contain the same random barcode.</p>
<p>The function <code>filterDuplicates</code> removes these PCR duplicate sequences (keeping only one copy in these cases), and creates de-duplicated BAM files. It returns the modified <code>CapSet</code> object with de-duplication statistics. <code>filterDuplicates</code> works in a strand-specific way. By default, it filters the R2 read from the paired-end BAM files, along with secondary alignments. Both reads can be allowed to kept using the option <code>keepPairs = TRUE</code>. Note that in this case, each read in a pair is evaluated seperately and in case of fragments where only one mate is flagged as duplicated, the other mate would be kept. In order to keep only fragments with both kept mates, the output BAM files should be filtered for proper pairs using samtools. Note that the run-time for paired-end evaluation is significantly longer, and only the first mate is considered for TSS detection (later).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load a previously saved CapSet object (or create new one)</span>
cs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/exampleCSobject.html">exampleCSobject</a></span>()
<span class="co">#&gt; Checking de-multiplexed R1 reads</span>
<span class="co">#&gt; Checking de-multiplexed R2 reads</span>
<span class="co">#&gt; Checking mapped file</span>
<span class="co">#&gt; Checking de-duplicated file</span>
<span class="co"># filter PCR duplicates and save output in a new directory</span>
<span class="kw">dir.create</span>(<span class="st">"removedup"</span>)
<span class="co">#&gt; Warning in dir.create("removedup"): 'removedup' already exists</span>
cs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/filterDuplicates.html">filterDuplicates</a></span>(cs, <span class="dt">outdir =</span> <span class="st">"removedup"</span>)
<span class="co">#&gt; Removing PCR duplicates : /home/bhardwaj/programs/svn/R-devel-build/library/icetea/extdata/bam/embryo1.bam</span>
<span class="co">#&gt; Removing PCR duplicates : /home/bhardwaj/programs/svn/R-devel-build/library/icetea/extdata/bam/embryo2.bam</span>
<span class="co">#&gt; Removing PCR duplicates : /home/bhardwaj/programs/svn/R-devel-build/library/icetea/extdata/bam/embryo3.bam</span>
<span class="co">#&gt; Removing PCR duplicates : /home/bhardwaj/programs/svn/R-devel-build/library/icetea/extdata/bam/embryo4.bam</span></code></pre></div>
<p>Tagging of read IDs using the random barcodes and subsequent removal of PCR duplicates can also be performed using the command line tools, like <a href="https://github.com/CGATOxford/UMI-tools">UMItools</a>.</p>
</div>
<div id="detection-of-tss" class="section level3">
<h3 class="hasAnchor">
<a href="#detection-of-tss" class="anchor"></a>Detection of TSS</h3>
<p><em>icetea</em> implements a new method of detection of transcription start sites, which is adopted from the recently described methods for differential transcription factor binding analysis in ChIP-Seq data <span class="citation">(Lun and Smyth 2015)</span>. Genome is divided into small, sliding windows and the TSSs are detected as the windows that show a user-defined fold-enrichment over a local background region. Multiple consicutively enriched windows are then merged to detect broader TSSs. The method works well with replicates, resulting in one TSS file per group.</p>
<p>This method is implemented in the function <code>detectTSS</code>, which returns a modified <code>CapSet</code> object with TSS detection statistics.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># detect TSS</span>
<span class="kw">dir.create</span>(<span class="st">"tssCalling"</span>)
<span class="co">#&gt; Warning in dir.create("tssCalling"): 'tssCalling' already exists</span>
cs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/detectTSS.html">detectTSS</a></span>(cs, <span class="dt">groups =</span> <span class="kw">c</span>(<span class="st">"wt"</span>, <span class="st">"wt"</span>, <span class="st">"mut"</span>, <span class="st">"mut"</span>), 
                 <span class="dt">outfile_prefix =</span> <span class="st">"tssCalling/testTSS"</span>, <span class="dt">restrictChr =</span> <span class="st">"X"</span>, <span class="dt">ncores =</span> <span class="dv">1</span>)
<span class="co">#&gt; Counting reads within detected TSS</span>
<span class="co">#&gt; Writing filtering information as .Rdata</span>
<span class="co"># export the detected TSS bed files</span>
<span class="kw"><a href="../reference/exportTSS.html">exportTSS</a></span>(cs, <span class="dt">merged =</span> <span class="ot">TRUE</span>, <span class="dt">outfile_prefix =</span> <span class="st">"tssCalling/testTSS"</span>)
<span class="co">#&gt; Writing merged .bed files</span></code></pre></div>
<p><a name="QC"></a></p>
</div>
</div>
<div id="plotting-and-qc" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-and-qc" class="anchor"></a>Plotting and QC</h2>
<p>The <code>sampleInfo</code> field of the CapSet object stores information about the read numbers kept at each step of processing, this information can be easily plotted using the function <code>plotReadStats</code></p>
<p>We can either plot the number of reads at each step of processing, or the proportion of reads w.r.t total demultiplexed reads per sample. Stacked or separate barplots can be made for each category.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># separated barchart for numbers</span>
<span class="kw"><a href="../reference/plotReadStats.html">plotReadStats</a></span>(cs, <span class="dt">plotValue =</span> <span class="st">"numbers"</span>, <span class="dt">plotType =</span> <span class="st">"dodge"</span>)
<span class="co">#&gt; </span></code></pre></div>
<p><img src="mapcap_analysis_files/figure-html/plotstats-1.png" width="672"></p>
<pre><code>#&gt; Plotting following information : samples demult_reads num_mapped num_filtered num_intss</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># stacked barchart for proportions</span>
<span class="kw"><a href="../reference/plotReadStats.html">plotReadStats</a></span>(cs, <span class="dt">plotValue =</span> <span class="st">"proportions"</span>, <span class="dt">plotType =</span> <span class="st">"stack"</span> )
<span class="co">#&gt; </span></code></pre></div>
<p><img src="mapcap_analysis_files/figure-html/plotstats2-1.png" width="672"></p>
<p>In case of well annotated genomes, one way to check the quality of TSS detection is to look at the fraction of detected TSSs that fall close to an annotated TSS in the genome. The cumulative fraction can be plotted for each sample, which can be used to compare samples. This can be done using the function <code>plotTSSprecision</code>, which takes the <em>known</em> TSS annotations as a <code>TxDB</code> object, along with either a <code>CapSet</code> object or a BED file, for TSS positions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"TxDb.Dmelanogaster.UCSC.dm6.ensGene"</span>)
<span class="co">#&gt; Loading required package: GenomicFeatures</span>
<span class="co">#&gt; Loading required package: BiocGenerics</span>
<span class="co">#&gt; Loading required package: parallel</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'BiocGenerics'</span>
<span class="co">#&gt; The following objects are masked from 'package:parallel':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span>
<span class="co">#&gt;     clusterExport, clusterMap, parApply, parCapply, parLapply,</span>
<span class="co">#&gt;     parLapplyLB, parRapply, parSapply, parSapplyLB</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     IQR, mad, sd, var, xtabs</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     anyDuplicated, append, as.data.frame, basename, cbind,</span>
<span class="co">#&gt;     colMeans, colnames, colSums, dirname, do.call, duplicated,</span>
<span class="co">#&gt;     eval, evalq, Filter, Find, get, grep, grepl, intersect,</span>
<span class="co">#&gt;     is.unsorted, lapply, lengths, Map, mapply, match, mget, order,</span>
<span class="co">#&gt;     paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind,</span>
<span class="co">#&gt;     Reduce, rowMeans, rownames, rowSums, sapply, setdiff, sort,</span>
<span class="co">#&gt;     table, tapply, union, unique, unsplit, which, which.max,</span>
<span class="co">#&gt;     which.min</span>
<span class="co">#&gt; Loading required package: S4Vectors</span>
<span class="co">#&gt; Loading required package: stats4</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'S4Vectors'</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     expand.grid</span>
<span class="co">#&gt; Loading required package: IRanges</span>
<span class="co">#&gt; Loading required package: GenomeInfoDb</span>
<span class="co">#&gt; Loading required package: GenomicRanges</span>
<span class="co">#&gt; Loading required package: AnnotationDbi</span>
<span class="co">#&gt; Loading required package: Biobase</span>
<span class="co">#&gt; Welcome to Bioconductor</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     Vignettes contain introductory material; view with</span>
<span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="kw">seqlevelsStyle</span>(TxDb.Dmelanogaster.UCSC.dm6.ensGene) &lt;-<span class="st"> "ENSEMBL"</span>
<span class="co"># only analyse genes on chrX to make the analysis faster</span>
<span class="kw">seqlevels</span>(TxDb.Dmelanogaster.UCSC.dm6.ensGene) &lt;-<span class="st"> "X"</span>
transcripts &lt;-<span class="st"> </span><span class="kw">transcripts</span>(TxDb.Dmelanogaster.UCSC.dm6.ensGene)

<span class="co"># Plotting the precision using a pre computed set of TSS (.bed files) :</span>
tssfile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"testTSS_merged.bed"</span>, <span class="dt">package =</span> <span class="st">"icetea"</span>)
<span class="kw"><a href="../reference/plotTSSprecision.html">plotTSSprecision</a></span>(<span class="dt">reference =</span> transcripts, <span class="dt">detectedTSS =</span> tssfile, <span class="dt">sampleNames =</span> <span class="st">"testsample"</span>)
<span class="co">#&gt; There are 13 regions with distance &gt; 500 bp to the closest TSS. They are all being reduced to 500 bp for the calculation. Samplewise numbers are : 13</span></code></pre></div>
<p><img src="mapcap_analysis_files/figure-html/plotPrecision-1.png" width="672"></p>
</div>
<div id="differential-tss-expression-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-tss-expression-analysis" class="anchor"></a>Differential TSS expression analysis</h2>
<p>For the experiments with two or more groups, <em>icetea</em> can be used to perform differential TSS expression analysis between groups. The requirements for differential TSS expression analysis is the same as that for differential expression analysis of RNA-Seq data. At least two or more biological replicates per group is required.</p>
<p>The functions <code>fitDiffTSS</code> and <code>detectDiffTSS</code> utilize <a href="http://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a> <span class="citation">(Robinson, McCarthy, and Smyth 2010)</span> to perform differential TSS expression analysis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## fitDiffTSS returns a DGEGLM object
csfit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fitDiffTSS.html">fitDiffTSS</a></span>(cs, <span class="dt">groups =</span> <span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">"wt"</span>,<span class="st">"mut"</span>), <span class="dt">each =</span> <span class="dv">2</span>), <span class="dt">normalization =</span> <span class="st">"windowTMM"</span>,
                     <span class="dt">outplots =</span> <span class="ot">NULL</span>, <span class="dt">plotref =</span> <span class="st">"embryo1"</span>)
<span class="kw">save</span>(csfit, <span class="dt">file =</span> <span class="st">"diffTSS_fit.Rdata"</span>)

## This object is then used by the detectDiffTSS function to return differentially expressed TSSs
de_tss &lt;-<span class="st"> </span><span class="kw"><a href="../reference/detectDiffTSS.html">detectDiffTSS</a></span>(csfit, <span class="dt">testGroup =</span> <span class="st">"mut"</span>, <span class="dt">contGroup =</span> <span class="st">"wt"</span>,
               <span class="dt">TSSfile =</span> <span class="kw">file.path</span>(dir, <span class="st">"testTSS_merged.bed"</span>), <span class="dt">MAplot_fdr =</span> <span class="fl">0.05</span>)
## export the output
<span class="kw">library</span>(rtracklayer)
<span class="kw"><a href="http://www.rdocumentation.org/packages/rtracklayer/topics/BEDFile-class">export.bed</a></span>(de_tss[de_tss<span class="op">$</span>score <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>], <span class="dt">con =</span> <span class="st">"diffTSS_output.bed"</span>)</code></pre></div>
<p><strong>Result:</strong> <code>de_tss</code> is a <code>GRanges</code> object that contain all the tested TSSs, along with metadata columns <code>score</code> (containing adjusted P-values) <code>logFC</code> (containing log-fold changes of group ‘test’ over group ‘control’) and logCPM (containing average expression of the TSS across all groups). This file can be exported to a bed file using the function <code>export.bed</code> from the package <code>rtracklayer</code></p>
<div id="using-spike-in-controls" class="section level3">
<h3 class="hasAnchor">
<a href="#using-spike-in-controls" class="anchor"></a>Using spike-In controls</h3>
<p>For the differential TSS expression analysis, the function <code>fitDiffTSS</code> shown above can utilize one of the available internal normalization methods. In some cases however, external (spike-in) normalizations are preferred. <em>icetea</em> provides a way to perform spike-in normalization during differential expression analysis.</p>
<p>Spike-In samples could be processed in the same way as the normal samples, using the <code>CapSet</code> object as described above. Normalization factors can be obtained from the spike-in reads using <code>calcNormFactors</code> and provided to the <code>fitDiffTSS</code> function during detection of differential TSS between samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## get gene counts for spike-in RNA mapped to human genome
<span class="kw">library</span>(<span class="st">"TxDb.Hsapiens.UCSC.hg38.knownGene"</span>)
normfacs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/calcNormFactors.html">calcNormFactors</a></span>(cs_spike, <span class="dt">features =</span> <span class="kw">genes</span>(TxDb.Hsapiens.UCSC.hg38.knownGene))

csfit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fitDiffTSS.html">fitDiffTSS</a></span>(cs, <span class="dt">groups =</span> <span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">"wt"</span>,<span class="st">"mut"</span>), <span class="dt">each =</span> <span class="dv">2</span>), 
                    <span class="dt">normalization =</span> <span class="ot">NULL</span>, <span class="dt">normFactors =</span> normfacs,
                    <span class="dt">outplots =</span> <span class="ot">NULL</span>, <span class="dt">plotRefSample =</span> <span class="st">"embryo1"</span>, <span class="dt">ncores =</span> <span class="dv">1</span>)</code></pre></div>
</div>
</div>
<div id="additional-useful-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#additional-useful-functions" class="anchor"></a>Additional useful functions</h2>
<div id="getting-the-gene-counts" class="section level3">
<h3 class="hasAnchor">
<a href="#getting-the-gene-counts" class="anchor"></a>Getting the gene counts</h3>
<p>Given the annotations, TSS counts for each gene can be summarized to gene counts. The following function sums up the TSS counts per gene from a given <code>txdb</code> object and returns gene counts.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get transcripts</span>
dm6trans &lt;-<span class="st"> </span><span class="kw">transcriptsBy</span>(TxDb.Dmelanogaster.UCSC.dm6.ensGene, <span class="st">"gene"</span>)
<span class="co"># get gene counts, counting reads around 500 bp of the TSS</span>
gcounts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getGeneCounts.html">getGeneCounts</a></span>(cs, dm6trans)</code></pre></div>
</div>
<div id="annotating-the-tss-distribution" class="section level3">
<h3 class="hasAnchor">
<a href="#annotating-the-tss-distribution" class="anchor"></a>Annotating the TSS distribution</h3>
<p>Detected or differentially expressed TSS, which are exported as bed files, can be quickly annotated using the function <code>annotateTSS</code>. <code>annotateTSS</code> reports the numbers/proportions of detected TSS falling into different genomic features (provided by a TxDb object) as a data.frame and optionally as a plot. In order to break ties between overlapping features, we can provide a vector of feature names in the decreasing order of preference.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># save the output as data.frame (outdf) + plot on screen</span>
outdf &lt;-<span class="st"> </span><span class="kw"><a href="../reference/annotateTSS.html">annotateTSS</a></span>(<span class="dt">tssBED =</span> tssfile,
            <span class="dt">txdb =</span> TxDb.Dmelanogaster.UCSC.dm6.ensGene,
            <span class="dt">plotValue =</span> <span class="st">"number"</span>,
            <span class="dt">outFile =</span> <span class="ot">NULL</span>)
<span class="co">#&gt; 'select()' returned 1:1 mapping between keys and columns</span>
<span class="co">#&gt; 'select()' returned many:1 mapping between keys and columns</span>
<span class="co">#&gt; 'select()' returned 1:1 mapping between keys and columns</span>
<span class="co">#&gt; 'select()' returned many:1 mapping between keys and columns</span>
<span class="co">#&gt; 'select()' returned many:1 mapping between keys and columns</span>
<span class="co">#&gt; 'select()' returned 1:1 mapping between keys and columns</span></code></pre></div>
<p><img src="mapcap_analysis_files/figure-html/annotateTSS-1.png" width="672"></p>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-batut2013">
<p>Batut, Philippe, Alexander Dobin, Charles Plessy, Piero Carninci, and Thomas R Gingeras. 2013. “High-Fidelity Promoter Profiling Reveals Widespread Alternative Promoter Usage and Transposon-Driven Developmental Gene Expression.” <em>Genome Research</em> 23 (1). Cold Spring Harbor Lab: 169–80.</p>
</div>
<div id="ref-kodzius2006cage">
<p>Kodzius, Rimantas, Miki Kojima, Hiromi Nishiyori, Mari Nakamura, Shiro Fukuda, Michihira Tagami, Daisuke Sasaki, et al. 2006. “CAGE: Cap Analysis of Gene Expression.” <em>Nature Methods</em> 3 (3). Nature Publishing Group: 211.</p>
</div>
<div id="ref-liao2013subread">
<p>Liao, Yang, Gordon K Smyth, and Wei Shi. 2013. “The Subread Aligner: Fast, Accurate and Scalable Read Mapping by Seed-and-Vote.” <em>Nucleic Acids Research</em> 41 (10). Oxford University Press: e108–e108.</p>
</div>
<div id="ref-lun2015csaw">
<p>Lun, Aaron TL, and Gordon K Smyth. 2015. “Csaw: A Bioconductor Package for Differential Binding Analysis of Chip-Seq Data Using Sliding Windows.” <em>Nucleic Acids Research</em> 44 (5). Oxford University Press: e45–e45.</p>
</div>
<div id="ref-robinson2010edger">
<p>Robinson, Mark D, Davis J McCarthy, and Gordon K Smyth. 2010. “EdgeR: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.” <em>Bioinformatics</em> 26 (1). Oxford University Press: 139–40.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#quick-start">Quick Start</a></li>
      <li><a href="#help-and-citations">Help and citations</a></li>
      <li><a href="#the-capset-object">The CapSet object</a></li>
      <li><a href="#analysis-workflow">Analysis workflow</a></li>
      <li><a href="#plotting-and-qc">Plotting and QC</a></li>
      <li><a href="#differential-tss-expression-analysis">Differential TSS expression analysis</a></li>
      <li><a href="#additional-useful-functions">Additional useful functions</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Vivek Bhardwaj.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
