<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Detect differentially expressed Transcription Start Sites between two conditions (fit model) — fitDiffTSS • icetea</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
<meta property="og:title" content="Detect differentially expressed Transcription Start Sites between two conditions (fit model) — fitDiffTSS" />

<meta property="og:description" content="Detect differentially expressed Transcription Start Sites between two conditions (fit model)" />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->




  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">icetea</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/mapcap_analysis.html">Analysing transcript 5'-profiling data using icetea</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/vivekbhr/icetea">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Detect differentially expressed Transcription Start Sites between two conditions (fit model)</h1>
    <small>Source: <a href='https://github.com/vivekbhr/icetea/blob/master/R/AllGenerics.R'><code>R/AllGenerics.R</code></a>, <a href='https://github.com/vivekbhr/icetea/blob/master/R/detectDiffTSS.R'><code>R/detectDiffTSS.R</code></a></small>
    <div class="hidden name"><code>fitDiffTSS.Rd</code></div>
    </div>

    
    <p>Detect differentially expressed Transcription Start Sites between two conditions (fit model)</p>
    

    <pre class="usage"><span class='fu'>fitDiffTSS</span>(<span class='no'>CSobject</span>, <span class='kw'>TSSfile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>groups</span>, <span class='kw'>normalization</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>normFactors</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>outplots</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>plotRefSample</span>, <span class='kw'>ncores</span> <span class='kw'>=</span> <span class='fl'>1</span>)

<span class='co'># S4 method for CapSet</span>
<span class='fu'>fitDiffTSS</span>(<span class='no'>CSobject</span>, <span class='kw'>TSSfile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>groups</span>,
  <span class='kw'>normalization</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>normFactors</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>outplots</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>plotRefSample</span>, <span class='kw'>ncores</span> <span class='kw'>=</span> <span class='fl'>1</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>CSobject</th>
      <td><p>An object of class <code><a href='newCapSet.html'>CapSet</a></code></p></td>
    </tr>
    <tr>
      <th>TSSfile</th>
      <td><p>A .bed file with TSS positions to test for differential TSS analysis. If left
empty, the union of detected TSS present within the provided CSobject would be plotted.</p></td>
    </tr>
    <tr>
      <th>groups</th>
      <td><p>Character vector indicating the group into which each sample within the CSobject falls.
the groups would be use to create a design matrix. As an example, replicates for one
condition could be in the same group.</p></td>
    </tr>
    <tr>
      <th>normalization</th>
      <td><p>A character indicating the type of normalization to perform . Options are "windowTMM",
"globalTMM" or NULL (don't compute normalization factors).
If "windowTMM" is chosen, the normalization factors are calculated using the TMM
method on 10 kb windows of the genome. "globalTMM" computes TMM normalization using counts
from all the evaluated TSSs. If NULL, the external normalization factors can be used
 (provided using `normFactors`).</p></td>
    </tr>
    <tr>
      <th>normFactors</th>
      <td><p>external normalization factors (from Spike-Ins, for example).</p></td>
    </tr>
    <tr>
      <th>outplots</th>
      <td><p>Output pdf filename for plots. If provided, the plots for BCV, dispersion and
MDS plot is created and saved in this file.</p></td>
    </tr>
    <tr>
      <th>plotRefSample</th>
      <td><p>Name of reference sample to plot for detection of composition bias in the
data. Samples could be normalized using one of the provided normalization methods to
control for composition bias.</p></td>
    </tr>
    <tr>
      <th>ncores</th>
      <td><p>No. of cores/threads to use</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An object of class <a href='http://www.rdocumentation.org/packages/edgeR/topics/DGEGLM-class'>DGEGLM-class</a>.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'># before running this
# 1. Create a CapSet object
# 2. de-multiplex the fastqs
# 3. map them
# 4. filter duplicate reads from mapped BAM
# 5. detect TSS
# 6. fit the diffTSS model
</div><span class='co'># NOT RUN {</span>
<span class='co'># load a previously saved CapSet object</span>
<span class='no'>cs</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='exampleCSobject.html'>exampleCSobject</a></span>()

<span class='co'># count reads on all TSS (union) and fit a model using replicates within groups</span>
<span class='no'>csfit</span> <span class='kw'>&lt;-</span> <span class='fu'>fitDiffTSS</span>(<span class='no'>cs</span>, <span class='kw'>groups</span> <span class='kw'>=</span> <span class='fu'>rep</span>(<span class='fu'>c</span>(<span class='st'>"wt"</span>,<span class='st'>"mut"</span>), <span class='kw'>each</span> <span class='kw'>=</span> <span class='fl'>2</span>), <span class='kw'>normalization</span> <span class='kw'>=</span> <span class='st'>"internal"</span>,
                     <span class='kw'>outplots</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>plotRefSample</span> <span class='kw'>=</span> <span class='st'>"embryo1"</span>)
<span class='fu'>save</span>(<span class='no'>csfit</span>, <span class='kw'>file</span> <span class='kw'>=</span> <span class='st'>"diffTSS_fit.Rdata"</span>)
<span class='co'># }</span><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Vivek Bhardwaj.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

