<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ icetea</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
<meta property="og:title" content="Function reference" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->




  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">icetea</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/mapcap_analysis.html">Analysing transcript 5'-profiling data using icetea</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/vivekbhr/icetea">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        <!--  -->
        <td>
          <p><code><a href="ResizeReads.html">ResizeReads()</a></code> </p>
        </td>
        <td><p>preprocess reads to count only 5' overlaps</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="activeChrs.html">activeChrs()</a></code> </p>
        </td>
        <td><p>Match BAM headers bw files and get active chromosome list (from restrict)
(written by Aaron Lun, 12 Dec 2014, copied and modified here)</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="annotateTSS.html">annotateTSS()</a></code> </p>
        </td>
        <td><p>Annotate the provided Transcription Start Sites</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="check_capSet.html">check_capSet()</a></code> </p>
        </td>
        <td><p>Check capset validity</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="demultiplexFASTQ.html">demultiplexFASTQ()</a></code> </p>
        </td>
        <td><p>Demultiplex and tag fastq files using sample barcodes</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="detectDiffTSS.html">detectDiffTSS()</a></code> </p>
        </td>
        <td><p>Detect differentially expressed Transcription Start Sites between two conditions (test)</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="detectTSS.html">detectTSS()</a></code> </p>
        </td>
        <td><p>Detection of Trancription start sites based on local enrichment</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="exampleCSobject.html">exampleCSobject()</a></code> </p>
        </td>
        <td><p>Create example CapSet object</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="exportTSS.html">exportTSS()</a></code> </p>
        </td>
        <td><p>Export the detected TSS from CapSet object as .bed files</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="filterDuplicates.html">filterDuplicates()</a></code> </p>
        </td>
        <td><p>Filter PCR-duplicates from mapped files using internal UMIs</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="filterDups.html">filterDups()</a></code> </p>
        </td>
        <td><p>Filter PCR-duplicates from BAM file using internal UMIs</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="fitDiffTSS.html">fitDiffTSS()</a></code> </p>
        </td>
        <td><p>Detect differentially expressed Transcription Start Sites between two conditions (fit model)</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="getBamFlags.html">getBamFlags()</a></code> </p>
        </td>
        <td><p>Get flags to read from bam</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="getChromBins.html">getChromBins()</a></code> </p>
        </td>
        <td><p>Get chromosome bins from BAM files</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="getChromWindows.html">getChromWindows()</a></code> </p>
        </td>
        <td><p>Get chromosome sliding windows from BAM files</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="getGeneCounts.html">getGeneCounts()</a></code> </p>
        </td>
        <td><p>Get gene-level counts from TSS data</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="getMCparams.html">getMCparams()</a></code> </p>
        </td>
        <td><p>Get platform-specific multicore params</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="getNormFactors.html">getNormFactors()</a></code> </p>
        </td>
        <td><p>Calculate normalization factors from CapSet object</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_newfastq.html">get_newfastq()</a></code> </p>
        </td>
        <td><p>Get data to create new ShortReadQ object after barcode trimming</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="getranks.html">getranks()</a></code> </p>
        </td>
        <td><p>Assign feature ranks on a VariantAnnotation output</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="mapCaps.html">mapCaps()</a></code> </p>
        </td>
        <td><p>Map the data from 5' profiling techniques</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="newCapSet.html">newCapSet()</a></code> </p>
        </td>
        <td><p>Create a new CapSet object</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="numReadsInBed.html">numReadsInBed()</a></code> </p>
        </td>
        <td><p>Count the number of reads in a given GRanges</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotPrecision.html">plotPrecision()</a></code> </p>
        </td>
        <td><p>Plotprecision background script</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotReadStats.html">plotReadStats()</a></code> </p>
        </td>
        <td><p>Plot read statistics from the CapSet object</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotTSSprecision.html">plotTSSprecision()</a></code> </p>
        </td>
        <td><p>Compare the precision of TSS detection between multiple samples</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="sampleInfo.html">sampleInfo()</a></code> <code><a href="sampleInfo.html">`sampleInfo<-`()</a></code> </p>
        </td>
        <td><p>Retrieve and replace sample information of a CapSet object</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="splitBAM_byIndex.html">splitBAM_byIndex()</a></code> </p>
        </td>
        <td><p>Split the composite BAM file using internal indexes (MAPCap)</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="splitBAM_byRepindex.html">splitBAM_byRepindex()</a></code> </p>
        </td>
        <td><p>Split the composite BAM file using replicate indexes (MAPCap data)</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="split_fastq.html">split_fastq()</a></code> </p>
        </td>
        <td><p>Split paired-end fastq by barcodes</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="splitranks.html">splitranks()</a></code> </p>
        </td>
        <td><p>Get features with the best rank for each TSS</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="strandBinCounts.html">strandBinCounts()</a></code> </p>
        </td>
        <td><p>Perform stranded Bin counts</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Vivek Bhardwaj.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

