<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Compare the precision of TSS detection between multiple samples — plotTSSprecision • icetea</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
<meta property="og:title" content="Compare the precision of TSS detection between multiple samples — plotTSSprecision" />

<meta property="og:description" content="Plot precision of TSS detection from multiple samples (bed files) with respect to
a given reference annotation.
Plot precision of TSS detection from multiple samples present within a
CapSet object, with respect to a given reference annotation." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->




  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">icetea</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/mapcap_analysis.html">Analysing transcript 5'-profiling data using icetea</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/vivekbhr/icetea">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compare the precision of TSS detection between multiple samples</h1>
    <small>Source: <a href='https://github.com/vivekbhr/icetea/blob/master/R/AllGenerics.R'><code>R/AllGenerics.R</code></a>, <a href='https://github.com/vivekbhr/icetea/blob/master/R/plotting_functions.R'><code>R/plotting_functions.R</code></a></small>
    <div class="hidden name"><code>plotTSSprecision.Rd</code></div>
    </div>

    
    <p>Plot precision of TSS detection from multiple samples (bed files) with respect to
a given reference annotation.</p>
<p>Plot precision of TSS detection from multiple samples present within a
<code><a href='newCapSet.html'>CapSet</a></code> object, with respect to a given reference annotation.</p>
    

    <pre class="usage"><span class='fu'>plotTSSprecision</span>(<span class='no'>reference</span>, <span class='no'>detectedTSS</span>, <span class='kw'>distanceCutoff</span> <span class='kw'>=</span> <span class='fl'>500</span>,
  <span class='kw'>outFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>...</span>)

<span class='co'># S4 method for GRanges,character</span>
<span class='fu'>plotTSSprecision</span>(<span class='no'>reference</span>, <span class='no'>detectedTSS</span>,
  <span class='kw'>distanceCutoff</span> <span class='kw'>=</span> <span class='fl'>500</span>, <span class='kw'>outFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>sampleNames</span>)

<span class='co'># S4 method for GRanges,CapSet</span>
<span class='fu'>plotTSSprecision</span>(<span class='no'>reference</span>, <span class='no'>detectedTSS</span>,
  <span class='kw'>distanceCutoff</span> <span class='kw'>=</span> <span class='fl'>500</span>, <span class='kw'>outFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>reference</th>
      <td><p>Reference Transcrips/Genes as a <code>GRanges</code> object</p></td>
    </tr>
    <tr>
      <th>detectedTSS</th>
      <td><p>Either a CapSet object with TSS information (after running <code><a href='detectTSS.html'>detectTSS</a></code>
or a character vector with paths to the BED files containing detcted TSSs</p></td>
    </tr>
    <tr>
      <th>distanceCutoff</th>
      <td><p>Maximum distance (in base pairs) from reference TSS to plot</p></td>
    </tr>
    <tr>
      <th>outFile</th>
      <td><p>Output file name (filename extention would be used to determine type)
If outfile not specified, the plot would be returned on the screen</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments</p></td>
    </tr>
    <tr>
      <th>sampleNames</th>
      <td><p>Labels for input samples (in the same order as the input bed files)</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A ggplot object, or a file. Plot showing perent of TSS detected per sample with respect to
        their cumulative distance to TSS of the provided reference</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='co'># load a txdb object</span>
<span class='fu'>library</span>(<span class='st'>"TxDb.Dmelanogaster.UCSC.dm6.ensGene"</span>)
<span class='fu'>seqlevelsStyle</span>(<span class='no'>TxDb.Dmelanogaster.UCSC.dm6.ensGene</span>) <span class='kw'>&lt;-</span> <span class='st'>"ENSEMBL"</span>
<span class='no'>transcripts</span> <span class='kw'>&lt;-</span> <span class='fu'>transcripts</span>(<span class='no'>TxDb.Dmelanogaster.UCSC.dm6.ensGene</span>)

<span class='co'># Plotting the precision using a pre computed set of TSS (.bed files) :</span>

<span class='no'>tssfile</span> <span class='kw'>&lt;-</span> <span class='fu'>system.file</span>(<span class='st'>"extdata"</span>, <span class='st'>"testTSS_merged.bed"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"icetea"</span>)
<span class='fu'>plotTSSprecision</span>(<span class='kw'>reference</span> <span class='kw'>=</span> <span class='no'>transcripts</span>, <span class='kw'>detectedTSS</span> <span class='kw'>=</span> <span class='no'>tssfile</span>,
                <span class='kw'>sampleNames</span> <span class='kw'>=</span> <span class='st'>"testTSS"</span>, <span class='kw'>distanceCutoff</span> <span class='kw'>=</span> <span class='fl'>500</span>,
                <span class='kw'>outFile</span> <span class='kw'>=</span> <span class='st'>"TSS_detection_precision.png"</span>)</div><div class='output co'>#&gt; <span class='message'>There are 13 regions with distance &gt; 500 bp to the closest TSS. They are all being reduced to 500 bp for the calculation. Samplewise numbers are : 13</span></div><div class='output co'>#&gt; <span class='message'>Saving 10.1 x 5.39 in image</span></div><div class='input'>
<span class='co'>## Plotting the precision using a CapSet object :</span>

<span class='fu'>library</span>(<span class='st'>"TxDb.Dmelanogaster.UCSC.dm6.ensGene"</span>)
<span class='fu'>seqlevelsStyle</span>(<span class='no'>TxDb.Dmelanogaster.UCSC.dm6.ensGene</span>) <span class='kw'>&lt;-</span> <span class='st'>"ENSEMBL"</span>
<span class='co'># only use chrX to make the analysis faster</span>
<span class='fu'>seqlevels</span>(<span class='no'>TxDb.Dmelanogaster.UCSC.dm6.ensGene</span>) <span class='kw'>&lt;-</span> <span class='st'>"X"</span>
<span class='no'>transcripts</span> <span class='kw'>&lt;-</span> <span class='fu'>transcripts</span>(<span class='no'>TxDb.Dmelanogaster.UCSC.dm6.ensGene</span>)

<span class='co'># load a previously saved CapSet object</span>
<span class='no'>cs</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='exampleCSobject.html'>exampleCSobject</a></span>()</div><div class='output co'>#&gt; <span class='message'>Checking de-multiplexed R1 reads</span></div><div class='output co'>#&gt; <span class='message'>Checking de-multiplexed R2 reads</span></div><div class='output co'>#&gt; <span class='message'>Checking mapped file</span></div><div class='output co'>#&gt; <span class='message'>Checking de-duplicated file</span></div><div class='input'><span class='co'># plot</span>
<span class='fu'>plotTSSprecision</span>(<span class='kw'>reference</span> <span class='kw'>=</span> <span class='no'>transcripts</span>, <span class='kw'>detectedTSS</span> <span class='kw'>=</span> <span class='no'>cs</span>,
                  <span class='kw'>outFile</span> <span class='kw'>=</span> <span class='st'>"TSS_detection_precision.png"</span>)</div><div class='output co'>#&gt; <span class='message'>There are 0 regions with distance &gt; 500 bp to the closest TSS. They are all being reduced to 500 bp for the calculation. Samplewise numbers are : </span></div><div class='output co'>#&gt; <span class='message'>Saving 10.1 x 5.39 in image</span></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Vivek Bhardwaj.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

